libvast/builtins/operators/pseudonymize.cpp

```diff
@@ -0,0 +1,182 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2022 The VAST Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <vast/address.hpp>
+#include <vast/arrow_table_slice.hpp>
+#include <vast/arrow_table_slice_builder.hpp>
+#include <vast/concept/convertible/data.hpp>
+#include <vast/concept/convertible/to.hpp>
+#include <vast/concept/parseable/to.hpp>
+#include <vast/concept/parseable/vast/data.hpp>
+#include <vast/pipeline_operator.hpp>
+#include <vast/plugin.hpp>
+#include <vast/type.hpp>
+
+#include <arrow/table.h>
+
+namespace vast::plugins::pseudonymize {
+
+/// The configuration of the pseudonymize pipeline operator.
+struct configuration {
+  std::string seed;
+  std::array<address::byte_type, vast::address::pseudonymization_seed_array_size>
+    seed_bytes{};
+  std::vector<std::string> fields;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, configuration& x) {
+    return f(x.seed, x.fields);
+  }
+
+  static inline const record_type& layout() noexcept {
+    static auto result = record_type{
+      {"seed", string_type{}},
+      {"fields", list_type{string_type{}}},
+    };
+    return result;
+  }
+};
+
+class pseudonymize_operator : public pipeline_operator {
+public:
+  pseudonymize_operator(configuration config) : config_{std::move(config)} {
+    parse_seed_string();
+  }
+
+  /// Applies the transformation to an Arrow Record Batch with a corresponding
+  /// VAST layout.
+  [[nodiscard]] caf::error
+  add(type layout, std::shared_ptr<arrow::RecordBatch> batch) override {
+    std::vector<indexed_transformation> transformations;
+    for (const auto& field : config_.fields) {
+      for (const auto& index : caf::get<record_type>(layout).resolve_key_suffix(
+             field, layout.name())) {
+        auto transformation = [&](struct record_type::field field,
+                                  std::shared_ptr<arrow::Array> array) noexcept
+          -> std::vector<std::pair<struct record_type::field,
+                                   std::shared_ptr<arrow::Array>>> {
+          if (!caf::holds_alternative<address_type>(field.type)) {
+            VAST_ASSERT(false
+                        && "record batch field to be pseudonymized but does "
+                           "not "
+                           "have address type");
+            VAST_WARN("Field {} is to be pseudonymized but does not contain "
+                      "IP "
+                      "address values; skipping pseudonymization",
+                      field);
+            return {{field, array}};
```

Comment:
Since this is a very local precondition violation (you should not create this transformation if `caf::holds_alternative<address_type>(caf::get<record_type>(layout).field(index).type)` is `false`), you can also just call `die` here instead of having the `if`, e.g., check before even creating the transformation whether the field you're pointing to is of the correct type.

NB: There's a two-argument overload to `VAST_ASSERT`, so you don't need to do `condition && "message"`, but rather can do `condition, "message"`.

Reply:
Should be fixed in 061a2d7c8

Reply:
Actually, I moved this check outside of the lambda in https://github.com/tenzir/vast/pull/2719/commits/d3d0d4aa4e3efd6a7a71c4d305d7c44c708bc9a6